File: DSA/ArrayHashing/ValidAnagram/valid_anagram.py

# https://leetcode.com/problems/valid-anagram/


""" 
1. Check that length of both strings matches

2. Count all occurances of each letter in both strings
    - Store in separate hashmap for each string, <char, count>

 3. Iterate over one of the hashmaps, check the count against
    the values in the other hashmap, must be exactly equal
"""


def valid_anagram(s, t):
    # Time: O(s + t)
    # Space: O(s + t)
    # return Counter(s) == Counter(t)
    # return sorted(s) == sorted(t) O(1) space complexity

    if len(s) != len(t):
        return False

    count_s = {}
    count_t = {}

    for i in range(len(s)):
        count_s[s[i]] = 1 + count_s.get(s[i], 0)
        count_t[t[i]] = 1 + count_t.get(t[i], 0)

    return True if count_s == count_t else False

    # for c in count_s:
    #     if count_s[c] != count_t.get(c, 0):
    #         return False

    # return True


if __name__ == '__main__':
    test1 = 'anagram'
    test2 = 'nagaram'
    test3 = 'apple'
    test4 = 'ppale'
    test5 = 'apple'
    test6 = 'ppalx'
    print(valid_anagram(test1, test2))
    print(valid_anagram(test3, test4))
    print(valid_anagram(test5, test6))

File: DSA_Problems/ArrayHashing/ValidAnagram/ValidAnagram.java
package DSA_Problems.ArrayHashing.ValidAnagram;

import java.util.HashMap;

// https://leetcode.com/problems/valid-anagram/

public class ValidAnagram {

    public static boolean findAnagram(String s1, String s2) {

        /*
         * Add both strings to hashmaps with counters for each character, compare both
         * hashmaps
         */
        if (s1.length() != s2.length())
            return false;

        HashMap<Character, Integer> charCount1 = new HashMap<>();
        HashMap<Character, Integer> charCount2 = new HashMap<>();

        for (char c : s1.toCharArray()) {
            charCount1.put(c, charCount1.getOrDefault(c, 0) + 1);
        }

        for (char c : s2.toCharArray()) {
            charCount2.put(c, charCount2.getOrDefault(c, 0) + 1);
        }

        return charCount1.equals(charCount2);

    }

    public static void main(String[] args) {

        System.out.println(findAnagram("anagram", "nagaram")); // true
        System.out.println(findAnagram("rat", "cat")); // false
        System.out.println(findAnagram("apple", "epalp")); // true
        System.out.println(findAnagram("anagram", "nagarax")); // false

    }
}
